### PR TITLE
Adds `d/tfe_scim_group` to fetch SCIM groups

### PR DESCRIPTION
**Pull request by skj-skj**
## Description
This PR adds `d/tfe_scim_group` to fetch a SCIM group synced from an IdP into Terraform Enterprise.

It looks up groups by `name`, performing an exact, case-insensitive match while filtering out fuzzy API results.

When the group is found, `id` is populated as SCIM group id.

This PR also adds acceptance tests. Because SCIM groups rely on SCIM tokens, and SCIM tokens share the SCIM settings singleton, the acceptance coverage is grouped into a single omnibus test to avoid race conditions.

It also updates the skip regex in `ci.yml` so SCIM-related tests are skipped in CI, since they depend on the same SCIM settings singleton.

Docs for `d/tfe_scim_group` are included as well.

The helper function `createSCIMGroup` has also been updated to handle a 429 error.

_Remember to:_

- [x] _Update the [Change Log](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/changelog-process.md)_
- [x] _Update the [Documentation](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/changelog-process.md#updating-the-documentation)_

## Testing plan

1. Spin up a TFE instance with SAML + SCIM enabled and run acceptance tests against it.

## External links

[JIRA](https://hashicorp.atlassian.net/browse/TF-35681)
[RFC](https://hermes-sharepoint.hashicorp.services/document/01XOO7K4IWFH44KZAPDJC3QXNKIRWEZMBA)

## Output from acceptance tests
1.  TESTARGS="-run TestAccTFESCIMGroupDataSource" ENABLE_TFE=1 envchain scim make testacc
    Untitled 7
2. [`Nightly TFE Tests #1179`](https://github.com/hashicorp/terraform-provider-tfe/actions/runs/26942550032)
    Untitled 6

## Rollback Plan

we can revert this pr

## Changes to Security Controls

NA

### CODE REVIEW

File: internal/provider/helper_test.go [hunk 1]
```diff
@@ -380,18 +380,39 @@ func createSCIMGroup(t *testing.T, displayName, scimToken string) string {
...
+		resp.Body.Close()
+		wait := time.Duration(1<<attempt) * time.Second
+		if ra := resp.Header.Get("Retry-After"); ra != "" {
+			if secs, err := time.ParseDuration(ra + "s"); err == nil {
+				wait = secs
+			}
+		}
+		t.Logf("SCIM rate-limited (429), retrying in %s (attempt %d/%d)", wait, attempt+1, maxRetries)
+		time.Sleep(wait)
 	}
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/helper_test.go:420]: fixed it, avoid reading closed body and wasted sleep on final SCIM 429 retry

File: internal/provider/helper_test.go [hunk 2]
```diff
@@ -380,18 +380,43 @@ func createSCIMGroup(t *testing.T, displayName, scimToken string) string {
...
+			if ra := resp.Header.Get("Retry-After"); ra != "" {
+				if secs, err := time.ParseDuration(ra + "s"); err == nil {
+					wait = secs
+				}
+			}
+			t.Logf("SCIM rate-limited (429), retrying in %s (attempt %d/%d)", wait, attempt+1, maxRetries)
+			time.Sleep(wait)
+		}
 	}
 	defer resp.Body.Close()
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/helper_test.go:422]: I have fixed it

File: internal/provider/data_source_scim_groups.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+		return
+	}
+
+	// ExactlyOneOf guarantees exactly one of `name` or `search` is set.
+	// If either is still unknown, defer to apply by returning without
+	// writing state; the framework will mark all computed attributes as
+	// unknown for downstream consumers.
+	var argument, value, query string
+	switch {
+	case data.Name.IsUnknown() || data.Search.IsUnknown():
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/data_source_scim_groups.go]: Not valid. Per the [Terraform data source docs](https://developer.hashicorp.com/terraform/language/data-sources#data-source-behavior) 

> Terraform may defer reading the data source to the apply phase when at least one of the arguments… cannot be predicted during the plan phase.

the framework marks outputs as `computed` automatically. By the time `Read` is invoked the inputs are always known, so the `IsUnknown()` guard is unreachable in practice.

File: internal/provider/data_source_scim_groups.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+		for _, g := range list.Items {
+			if argument == "name" && !strings.EqualFold(g.Name, value) {
+				continue
+			}
+			matched = append(matched, g)
+		}
+
+		if argument == "name" && len(matched) > 0 {
+			break
+		}
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/data_source_scim_groups.go]: not valid, as TFE backend ensure uniqueness in SCIM groups, so first exact match is also the last exact match

File: internal/provider/data_source_scim_groups.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+				Optional:    true,
+				Description: "A search string used to filter SCIM groups by name via the API's query parameter (`?q=<search>`, case-insensitive). Cannot be used with `name`.",
+				Validators: []validator.String{
+					stringvalidator.LengthAtLeast(1),
+					stringvalidator.RegexMatches(
+						scimGroupNonWhitespaceRegex,
+						"must contain at least one non-whitespace character",
+					),
+				},
+			},
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/data_source_scim_groups.go]: fixed it, added `path.MatchRelative().AtParent().AtName("name")` in search

**Comment by nikolasrieble** [comment]
Nice work on this data source! One suggestion for the test coverage:

**Consider extracting the exact-match filtering logic into a pure helper function and unit-testing it independently.**

The client-side `strings.EqualFold` filtering in `Read()` is the most important behavioral contract of the `name` mode (since the API only does substring matching). Right now it's only exercised through acceptance tests, which are slow and coupled to a live TFE instance.

The repo already has an established pattern for this — e.g. `workspace_helpers.go` / `workspace_helpers_test.go`, `id_helpers.go` / `id_helpers_test.go`. You could extract something like:

```go
// scim_groups_helpers.go
func filterExactSCIMGroups(groups []*tfe.AdminSCIMGroup, name string) []*tfe.AdminSCIMGroup {
    var matched []*tfe.AdminSCIMGroup
    for _, g := range groups {
        if strings.EqualFold(g.Name, name) {
            matched = append(matched, g)
        }
    }
    return matched
}
```

Then in `scim_groups_helpers_test.go` you can cheaply cover the edge cases (case-insensitive match, fuzzy siblings rejected, duplicate display names) with fast unit tests that don't need SCIM infrastructure. This also lets you simplify the acceptance test's `name` lifecycle — it wouldn't need separate steps for fuzzy filtering and case-insensitivity since those are covered at the unit level.

File: internal/provider/helper_test.go [hunk 6]
```diff
@@ -380,18 +380,43 @@ func createSCIMGroup(t *testing.T, displayName, scimToken string) string {
...
 	httpClient := &http.Client{Timeout: 30 * time.Second}
-	resp, err := httpClient.Do(req)
-	if err != nil {
-		t.Fatalf("POST /scim/v2/Groups: %v", err)
+
+	const maxRetries = 5
+	var resp *http.Response
+	for attempt := range maxRetries {
+		// Re-create the request body each attempt since the reader is consumed.
+		reqBody := bytes.NewReader(body)
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/helper_test.go:389]: actually, it does. from golang v1.22 [release notes](https://go.dev/doc/go1.22#language) 

> “For” loops may now range over integers.

File: internal/provider/helper_test.go [hunk 7]
```diff
@@ -380,18 +380,43 @@ func createSCIMGroup(t *testing.T, displayName, scimToken string) string {
...
-
 	httpClient := &http.Client{Timeout: 30 * time.Second}
-	resp, err := httpClient.Do(req)
-	if err != nil {
-		t.Fatalf("POST /scim/v2/Groups: %v", err)
+
+	const maxRetries = 5
+	var resp *http.Response
+	for attempt := range maxRetries {
+		// Re-create the request body each attempt since the reader is consumed.
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/helper_test.go:388]: the `go.mod` file contains `go 1.25.10` not `go 1.21`, check again.

**Comment by skj-skj** [comment]
@nikolasrieble added a helper function and its acc test

File: internal/provider/data_source_scim_groups.go [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+type modelDataTFESCIMGroup struct {
+	ID   types.String `tfsdk:"id"`
+	Name types.String `tfsdk:"name"`
+}
+
+// modelDataTFESCIMGroups maps the data source schema data.
+type modelDataTFESCIMGroups struct {
+	ID        types.String            `tfsdk:"id"`
+	Name      types.String            `tfsdk:"name"`
+	Search    types.String            `tfsdk:"search"`
```
Comment (kkavish) [inline internal/provider/data_source_scim_groups.go]: I wonder why even support fuzzy search here? My point is - if the only use-case is to link teams and update admin group in SCIM, operators would need the exact name anyways. We do not support updates/deletion on SCIM managed groups, so getting all groups matching a pattern and then doing an op on them is definitely not an use-case.

Could also simplify a lot of implementation and testing here.

**Review by kkavish — CHANGES_REQUESTED** [review]

File: internal/provider/data_source_scim_groups.go [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+	// unknown for downstream consumers.
+	var argument, value, query string
+	switch {
+	case data.Name.IsUnknown() || data.Search.IsUnknown():
+		return
+	case !data.Name.IsNull():
+		argument = "name"
+		value = strings.TrimSpace(data.Name.ValueString())
+		data.Name = types.StringValue(value)
+		query = value
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/data_source_scim_groups.go]: fixed

File: internal/provider/data_source_scim_groups.go [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+	case !data.Search.IsNull():
+		argument = "search"
+		value = strings.TrimSpace(data.Search.ValueString())
+		data.Search = types.StringValue(value)
+		query = value
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/data_source_scim_groups.go]: fixed

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -60,7 +60,7 @@ jobs:
           # worked reliably for a single test name and broke when excluding multiple suites. We now keep
           # list_tests at its default (".") and use the `skip` input below, which is forwarded to `go test -skip`
           # to reliably exclude multiple specific test suites from CI runs.
-          skip: "TestAccTFESAMLSettings_omnibus|TestAccTFESCIMSettings_omnibus|TestAccTFESCIMTokenDataSource_omnibus|TestAccTFESCIMToken_omnibus"
+          skip: "TestAccTFESAMLSettings_omnibus|TestAccTFESCIM"
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply .github/workflows/ci.yml:63]: every SCIM test, depends on `SCIM Settings` singleton, so this is valid.

File: internal/provider/data_source_scim_groups.go (hunk 8 above)
Comment (skj-skj) [reply internal/provider/data_source_scim_groups.go]: good point, agreed 👍. the exact `name` is required for the only real use case which is to link teams or site admin group. `search` (fuzzy matching)  doesn't add any value here. i have dropped the `search` argument and updated the test cases. thx for the suggestion!

File: internal/provider/data_source_scim_groups.go [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+				Required:    true,
+				Description: "The exact name of the SCIM group to retrieve (case-insensitive).",
+				Validators: []validator.String{
+					stringvalidator.LengthAtLeast(1),
+					stringvalidator.RegexMatches(
+						scimGroupNonWhitespaceRegex,
+						"must contain at least one non-whitespace character",
+					),
+				},
+			},
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/data_source_scim_groups.go]: updated pr description

File: internal/provider/helper_test.go [hunk 13]
```diff
@@ -380,18 +380,43 @@ func createSCIMGroup(t *testing.T, displayName, scimToken string) string {
...
+		if err != nil {
+			t.Fatalf("build SCIM group request: %v", err)
+		}
+		retryReq.Header.Set("Content-Type", "application/scim+json")
+		retryReq.Header.Set("Authorization", "Bearer "+scimToken)
+
+		resp, err = httpClient.Do(retryReq)
+		if err != nil {
+			t.Fatalf("POST /scim/v2/Groups: %v", err)
+		}
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/helper_test.go:401]: already closing the body here `defer resp.Body.Close()`

File: CHANGELOG.md [hunk 14]
```diff
@@ -4,6 +4,7 @@ FEATURES:
 * **New Resource:** `r/tfe_scim_settings` and **New Data Source:** `d/tfe_scim_settings`: Adds resource and data source to manage SCIM settings on Terraform Enterprise. By @skj-skj [#2072](https://github.com/hashicorp/terraform-provider-tfe/pull/2072)
 * **New Resource:** `r/tfe_scim_token` and **New Data Source:** `d/tfe_scim_token`: Adds resource and data source to manage SCIM tokens on Terraform Enterprise. By @skj-skj [#2076](https://github.com/hashicorp/terraform-provider-tfe/pull/2076)
 * **New Resource:** `r/tfe_admin_smtp_settings` and **New Data Source:** `d/tfe_admin_smtp_settings`: Adds resource and data source to manage Admin SMTP settings in Terraform Enterprise by @chpag [#2059](https://github.com/hashicorp/terraform-provider-tfe/pull/2059)
+* **New Data Source:** `d/tfe_scim_group`: Adds a data source to retrieve a SCIM group from Terraform Enterprise. By @skj-skj [#2083](https://github.com/hashicorp/terraform-provider-tfe/pull/2083)
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply CHANGELOG.md:7]: fixed the pr title

File: internal/provider/data_source_scim_group.go [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+	client, ok := req.ProviderData.(ConfiguredClient)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected tfe.ConfiguredClient, got %T. This is a bug in the tfe provider, so please report it on GitHub.", req.ProviderData),
+		)
+
+		return
+	}
+	d.client = client.Client
```
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/data_source_scim_group.go:89]: Leave as-is to match the convention

File: internal/provider/helper_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (skj-skj) [reply internal/provider/helper_test.go:401]: The `defer resp.Body.Close()` after the loop already handles this, it runs before every `t.Fatalf`, so the final response body is always closed.

**Review by kkavish — DISMISSED** [review]

File: internal/provider/helper_test.go [hunk 16]
```diff
@@ -380,18 +380,46 @@ func createSCIMGroup(t *testing.T, displayName, scimToken string) string {
...
+			fmt.Sprintf("https://%s/scim/v2/Groups", hostname), reqBody)
+		if err != nil {
+			t.Fatalf("build SCIM group request: %v", err)
+		}
+		retryReq.Header.Set("Content-Type", "application/scim+json")
+		retryReq.Header.Set("Authorization", "Bearer "+scimToken)
+
+		resp, err = httpClient.Do(retryReq)
+		if err != nil {
+			if resp != nil && resp.Body != nil {
```
Comment (shwetamurali) [inline internal/provider/helper_test.go]: Nit: when `err != nil`, `resp` is always nil. You can simplify to just `t.Fatalf(...)` directly.

**Review by shwetamurali — DISMISSED** [review]

File: internal/provider/helper_test.go (hunk 16 above)
Comment (skj-skj) [reply internal/provider/helper_test.go]: good catch, fixed

**Review by shwetamurali — APPROVED** [review]
